Model/DutyCalculator.php

```diff
@@ -116,84 +116,333 @@ public function __construct(
         $this->_logger = $logger;
     }
 
+
+    /**
+     * Sets quote and checkout Session state based on whether duty and tax is applicable or not
+     * @param Magento\Quote\Api\Data\AddressInterface $address
+     * @param bool $apply  - whether duty is applicable or not
+     *
+     */
+    public function handleTaxApplicability($address, $apply)
+    {
+        $this->_logger->debug('In handleTaxApplicability method');
+        $this->_logger->debug($this->checkoutSession->getReachDuty());
+        $this->_logger->debug($address->getCountryId());
+        $this->_logger->debug($address->getRegionCode());
+        $this->_logger->debug($this->checkoutSession->getPrevCountry());
+        $this->_logger->debug($this->checkoutSession->getPrevRegion());
+        $this->_logger->debug($apply);
+        $quote = $this->checkoutSession->getQuote();
+
+        if ($apply || !$this->getIsOptional($address->getCountryId())) {
+           $quote->setReachDuty($this->checkoutSession->getReachDuty());
+           $this->response->setIsOptional($this->getIsOptional($address->getCountryId()));
+           $this->response->setSuccess(true);
+           $this->response->setDuty($this->checkoutSession->getRechDuty());
+           $this->_logger->debug('In handleTaxApplicability method --- apply is true');
+        }
+        else {
+           $this->response->setIsOptional($this->getIsOptional($address->getCountryId()));
+           $this->response->setSuccess(true);
+           //as apply duty and tax(DT) is not selected duty and tax would not be used in total pricing
+           //so setting all relevant quote values to zero
+           //$quote->setDuty(0) ;
+           $quote->setReachDuty(0) ;
+           $this->response->setDuty($this->checkoutSession->getReachDuty()); //the DT checkbox and label  should still appear
+           //if this amount is more than 0 even though the user did not choose to apply it on
+           //computation of total landed cost/billing
+           $this->_logger->debug('In handleTaxApplicability method --- apply is false');
+        }
+
+        $quote->save();//this is needed so that different aspects related to a quote are available on other pages.
+        //More specifically saving quote in database is needed as apparently Magento session scopes are not application
+        // wide but code area specific (?)
```

Comment:
@ryanjohnston could you confirm this so that our comment is accurate and factual going forward

Reply:
To help with this a bit:
The information that we are manipulating in dutyCalculation.php is required in duty.php as well to show proper values of applicable duty. With the code as is that block is looking at quote info /state that is in the db not what I stored in session in dutyCalculation.php. To make sure proper state is available in duty.php this state save thing was done. Thanks.

Reply:
And the session was in dutyCalculation.php so that I do not have to make another call to retrieve duty from DHL. When user deselect the apply duty checkbox the duty is set to zero in db state but the value returned from DHL api call is saved in session to reuse it if the user changes mind or when that page is getting entered repeatedly due to different events (mostly Magento ones). Instead of session this could be save din database too. However that would need creation of new table and or fields in the database.

Reply:
https://mage2-blog.com/magento-2-get-session/ has some info on sessions. I should have a better link but not handy right now.

Adding checkout session in duty.php and then retrieving value from there could work too as another possible solution. But then do not know (without checking) where else that fix would break due to the state not being saved in db. Significant time was gone chasing events from all directions.

Reply:
Definitely! I appreciate how thorough you looked into this, it makes sense why we would need to save the quote every time and this background definitely helps. Do you think it would be helpful to fix this comment up with what you know about saving the quote / how the session and quote interact with the code as it exists right now?

My thinking is that I don't want to leave questions in the comments, I prefer to leave explanatory comments that give the reader more information that is directly relevant to the code around it. However, commenting styles are very individual/unique so I don't want to infringe on your personal style. Hopefully that comes across!

Reply:
Removed the question mark and updated the comment. https://github.com/withreach/magento-plugin/pull/44/commits/b4543f304265925b832162585e43be63bceff755 Thanks @lizzynibali

Reply:
Makes sense after reading everything Suraiya has put together regarding having to save into the session.